### PR TITLE
FINERACT-1391:AccountNumberPreferencesPrefixEight

### PR DESCRIPTION
**Pull request by rrpawar96**
## Description

Describe the changes made and why they were made.

Ignore if these details are present on the associated [Apache Fineract JIRA ticket](https://github.com/apache/fineract/pull/1284).

## Checklist

Please make sure these boxes are checked before submitting your pull request - thanks!

- [ ] Write the commit message as per https://github.com/apache/fineract/#pull-requests

- [ ] Acknowledge that we will not review PRs that are not passing the build _("green")_ - it is your responsibility to get a proposed PR to pass the build, not primarily the project's maintainers.

- [ ] Create/update unit or integration tests for verifying the changes made.

- [ ] Follow coding conventions at https://cwiki.apache.org/confluence/display/FINERACT/Coding+Conventions.

- [ ] Add required Swagger annotation and update API documentation at fineract-provider/src/main/resources/static/api-docs/apiLive.htm with details of any API changes

- [ ] Submission is not a "code dump".  (Large changes can be made "in repository" via a branch.  Ask on the developer mailing list for guidance, if required.)

FYI our guidelines for code reviews are at https://cwiki.apache.org/confluence/display/FINERACT/Code+Review+Guide.

### CODE REVIEW

**Comment by bharathcgowda** [comment]
@rrpawar96  thank you for sending this fix, could you please check and fix the travis issue? 

@ptuomola  @vidakovic @thesmallstar  It would be really helpful if any one of you do the code review.

**Comment by vidakovic** [comment]
btw: I have a fix for the Travis issue... will submit later

**Comment by rrpawar96** [comment]
Yes, it seems the Travis build is been fixed by this PR #1981, let me also try again by rebasing.

**Review by ptuomola — CHANGES_REQUESTED** [review]
@rrpawar96 this looks like a good feature - but why are we hardcoding the prefix to be 8's only? Would this not be much more flexible if the user could specify any string (numeric?) to be used as prefix?

**Comment by rrpawar96** [comment]
Hi @ptuomola, The comments has been addressed, please re-review the PR.
Two param digits and prefixCharacter, 'prefixCharacter' will be taken to assign the prefix which will be respective length which will stored in the 'digits' param

**Comment by francisguchie** [comment]
@ptuomola @rrpawar96  & @vidakovic 

I would like to do a functional test on this please advise

**Comment by ikimbra** [comment]
@ptuomola I see the issue has now been fixed.

**Comment by francisguchie** [comment]
@rrpawar96  @bharathcgowda  @ptuomola 
I pulled this your branch and did a functional test - It works well for me 

> @ptuomola I see the issue has now been fixed.

@ikimbra  did you do a test please advise

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/data/AccountNumberFormatDataValidator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/data/AccountNumberFormatDataValidator.java]: What's 401? Instead of using a "magic number" can we use the correct constant for this?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormat.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormat.java]: Same as above

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/service/AccountNumberFormatWritePlatformServiceJpaRepositoryImpl.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/service/AccountNumberFormatWritePlatformServiceJpaRepositoryImpl.java]: As above

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java]: Why do we need this block? We've already determined the prefix above....  can't we just use the length of the prefix to determine how much we need to pad the account number? Why do we need to retrieve the length again?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java]: Why do we need to look up again the AccountNumberFormat? Should we not use the one that we've been passed in as an argument, rather than looking up a new one?

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormat.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormat.java]: Why would we only set these variables for a specific type? What people expect is that the constructor will set the parameters that are passed in... what's the harm of just setting these every time?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java]: Do we need to pass in digits as a separate argument and store it separately in the database? Could we not infer the length of the prefix from the length of the prefix string? Ie if we want the prefix to be 3 characters of "123" we just store "123" as the prefix string.

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java]: This logic seems to be a) hardcoded to expect a specific length for prefix and digits b) only handle specific cases of length/prefix length. Can this not be made generic so that we automatically work out the length of the number / prefix based on the lengths of the inputs?

**Review by ptuomola — CHANGES_REQUESTED** [review]

**Comment by rrpawar96** [comment]
Thanks, @ptuomola for reviewing, I have addressed all comments, Requesting for re-review.
Removed the use of Digits and integer, used the length of Prefix and constant name instead, created new generic logic for creating Prefix and tested.

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormatRepository.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormatRepository.java]: Is this query still needed? I don't see it being called anywhere

File: fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormatRepository.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/infrastructure/accountnumberformat/domain/AccountNumberFormatRepository.java]: Same as above

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java]: Why is this change needed?

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java]: Same as above

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java]: Why do we need to initialise this again? It was initialise already in line 103...

File: fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java
Comment (ptuomola) [inline fineract-provider/src/main/java/org/apache/fineract/portfolio/client/domain/AccountNumberGenerator.java]: Should we call this function "generatePrefix" rather than "prefixEight"?

**Review by ptuomola — CHANGES_REQUESTED** [review]
Looks great @rrpawar96 - thanks for making the changes! A few more comments - hope you can address these, and then I think it should be ready for merging...

**Comment by rrpawar96** [comment]
@ptuomola Comment has been addressed, Removed unused properties.

**Review by ptuomola — APPROVED** [review]
LGTM... will wait until the build passes and then merge this

**Pull request merged**
